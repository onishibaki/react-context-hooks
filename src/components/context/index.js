import { createContext, useContext } from "react";

export const EmployeesContext = createContext();

export const useEmployeeState = () => useContext(EmployeesContext);
