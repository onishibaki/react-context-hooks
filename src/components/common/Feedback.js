import React, { useState } from "react";
import { useEmployeeState } from "../context";
import styled from "styled-components";

import {
  ActionTypes,
  updateFeedback,
  addFeedback,
  updateEmployeeFlag,
} from "../action";

const FeedBackContainer = styled.div`
  border: 1px solid black;
  background: rgba(0, 0, 0, 0.5);
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  width: 100%;
  height: 100%;
  text-align: center;
`;

const FeedBackWrapper = styled.div`
  border: 1px solid rgb(17, 17, 17);
  background-color: white;
  margin-top: 158px;
  display: inline-block;
  padding: 35px 10px 10px 10px;
  position: relative;
`;

const SpanButton = styled.span`
  font-weight: 700;
  color: black;
  font-size: 15px;
  margin: 10px 0;
  display: inline-block;
  position: absolute;
  top: -8px;
  right: 9px;
  cursor: pointer;
}
`;

const Button = styled.button`
  text-decoration: none;
  border: 1px solid #000000;
  padding: 10px 20px;
  font-weight: 700;
  color: black;
  font-size: 12px;
  margin: 10px 0;
  display: block;
  &:hover {
    background: #00000059;
    color: white;
  }
`;

const Textarea = styled.textarea`
  display: block;
  height: 100px;
  width: 350px;
`;

const Label = styled.label`
  display: block;
  text-align: left;
`;

export const Feedback = () => {
  //Use Context
  const [
    {
      formValue,
      employeeId,
      feedbackValue: { feedback_id: feedBackId, feedback },
      formMessage,
    },
    dispatch,
  ] = useEmployeeState();
  //Set state for FeedBack Changes
  const [stateFeedback, setFeedback] = useState("");

  //Checking for State if feedBack name is blank to throw error
  const [stateUpdateFeedback, setUpdateFeedback] = useState(null);

  //Set the First Value
  let feedBackParameter = feedback;

  //Parameter for Save
  const feedBackSaveParam = {
    employeeId,
    stateFeedback,
    flag: stateFeedback ? 0 : 1,
  };

  //Checking for State if feedBack is blank to throw error
  if (stateUpdateFeedback === "") {
    feedBackParameter = stateUpdateFeedback;
  }

  //Parameter for Update
  const feedBackUpdateParam = {
    employeeId,
    feedBackId,
    stateFeedback: stateFeedback || feedBackParameter,
  };

  //Getting the value of Input changes
  const inputChangedHandler = (event) => {
    setFeedback(event.target.value);
    setUpdateFeedback("");
  };

  return (
    //Condition to display Feedback Screen
    formValue !== "" && (
      <FeedBackContainer className="feedback-container">
        <FeedBackWrapper>
          <SpanButton
            onClick={() => {
              dispatch({
                type: ActionTypes.FORM_MESSAGE,
                data: "",
              });
              dispatch({
                type: ActionTypes.SET_FORM_STATE_FEEDBACK,
                form: "",
                data: {},
                id: "",
              });
              dispatch({
                type: ActionTypes.SET_FORM_STATE,
                form: "",
                data: {},
              });
              setFeedback("");
              setUpdateFeedback(null);
            }}
          >
            X
          </SpanButton>
          <Label>Feedback: </Label>
          <Textarea
            name="feedback"
            defaultValue={feedback}
            onChange={inputChangedHandler}
          ></Textarea>
          {Object.keys(formMessage).length !== 0 && (
            <h2>{formMessage.message}</h2>
          )}
          {formValue !== "update" ? (
            <Button
              onClick={() => {
                addFeedback(feedBackSaveParam).then(function (response) {
                  dispatch(response);
                });
                updateEmployeeFlag(feedBackSaveParam);
              }}
            >
              Save
            </Button>
          ) : (
            <Button
              onClick={() => {
                updateFeedback(feedBackUpdateParam).then(function (response) {
                  dispatch(response);
                });
              }}
            >
              Update
            </Button>
          )}
        </FeedBackWrapper>
      </FeedBackContainer>
    )
  );
};
