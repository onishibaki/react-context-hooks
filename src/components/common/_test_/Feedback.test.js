import React from "react";
import { render, cleanup } from "@testing-library/react";
import renderer from "react-test-renderer";
import Adapter from "enzyme-adapter-react-16";
import { configure, mount } from "enzyme";

import * as EmployeesContext from "../../context";
import { Feedback } from "../Feedback";

configure({ adapter: new Adapter() });

describe("Feedback Component Test", () => {
  let setValue;
  let container;
  beforeEach(() => {
    setValue = (dataParameter) => {
      jest
        .spyOn(EmployeesContext, "useEmployeeState")
        .mockImplementation(() => dataParameter);
    };
    container = document.createElement("div");
  });

  afterEach(cleanup);

  it("Feedback Component Renders Correctly", () => {
    const data = [
      {
        formValue: "save",
        feedbackValue: {},
        formMessage: "",
      },
    ];
    setValue(data);
    render(<Feedback />, container);
  });

  it("Check Feedback Save Button", () => {
    const data = [
      {
        formValue: "save",
        feedbackValue: {},
        formMessage: "",
      },
    ];
    setValue(data);
    const wrapper = mount(<Feedback />);
    const save = wrapper.find("button").text();
    expect(save).toEqual("Save");
  });

  it("Check Feedback Update Button", () => {
    const data = [
      {
        formValue: "update",
        feedbackValue: {},
        formMessage: "",
      },
    ];
    setValue(data);
    const wrapper = mount(<Feedback />);
    const update = wrapper.find("button").text();
    expect(update).toEqual("Update");
  });

  it("Check Error Message Displayed", () => {
    const data = [
      {
        formValue: "save",
        feedbackValue: {},
        formMessage: {
          status: 0,
          message: "Feedback is a required field",
        },
      },
    ];
    setValue(data);
    const wrapper = mount(<Feedback />);
    const h2 = wrapper.find("h2").text();
    expect(h2).toEqual("Feedback is a required field");
  });

  it("Check Textarea Display Feedback", () => {
    const data = [
      {
        formValue: "update",
        feedbackValue: {
          feedback_id: 1,
          feedback: "Nice Work",
          assign_id: 0,
        },
        formMessage: {},
      },
    ];
    setValue(data);
    const wrapper = mount(<Feedback />);
    const textarea = wrapper.find("textarea").text();
    expect(textarea).toEqual("Nice Work");
  });

  it("Check Feedback Component Test to Match Snapshot Display", () => {
    const data = [
      {
        formValue: "update",
        feedbackValue: {
          feedback_id: 1,
          feedback: "Nice Work",
          assign_id: 0,
        },
        formMessage: {},
      },
    ];
    setValue(data);
    const Snapshot = renderer.create(<Feedback />).toJSON();
    expect(Snapshot).toMatchSnapshot();
  });
});
