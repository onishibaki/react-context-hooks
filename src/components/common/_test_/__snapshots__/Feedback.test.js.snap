// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feedback Component Test Check Feedback Component Test to Match Snapshot Display 1`] = `
<div
  className="sc-AxjAm EmuRM feedback-container"
>
  <div
    className="sc-AxirZ ezqyFp"
  >
    <span
      className="sc-AxiKw qPNqo"
      onClick={[Function]}
    >
      X
    </span>
    <label
      className="sc-AxgMl cJEGFC"
    >
      Feedback: 
    </label>
    <textarea
      className="sc-AxhUy gpmBrv"
      defaultValue="Nice Work"
      name="feedback"
      onChange={[Function]}
    />
    <button
      className="sc-AxhCb bBarFh"
      onClick={[Function]}
    >
      Update
    </button>
  </div>
</div>
`;
